### ISSUE TITLE
If no supported_groups are available then we send an empty list

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#13583]
It is possible to not have any supported_groups available, and in that case a libssl client will send an empty list which is a protocol error (the supported_groups extension if sent must always have at least one element in it).

For example, if you configure OpenSSL with "no-ec2m" and then attempt to send a curve list containing only an ec2m curve: 
````
$ openssl s_client -no_tls1_3 -groups sect163k1
````

Then you get this error on the server:
````
$ openssl s_server
Using default temp DH parameters
ACCEPT
ERROR
140253113922880:error:1424206E:SSL routines:tls_parse_ctos_supported_groups:bad extension:ssl/statem/extensions_srvr.c:964:
shutting down SSL
CONNECTION CLOSED
````

Affects both master and 1.1.1

**Comment by t8m** [comment]
Hmm, interesting. What would be the correct behavior? Should we disable ECDH ciphersuites then? Or just fail on the client side?

**Comment by mattcaswell** [comment]
Maybe we could expand the current `use_ecc()` function to check for this and therefore disable EC usage if it fails?

**Comment by kaduk** [comment]
[This largely seems to only affect pre-TLS 1.3, since TLS 1.3 if you have no groups you can only do a PSK key exchange.]
Coming at it from the mindset that brought us TLS 1.3 would suggest to just fail on the client side, but I think it is probably better to do something like what Matt suggests.

**Comment by baentsch** [comment]
> Affects both master and 1.1.1

Confirmed for 1.1.1 but not for master (as of today): There, `./Configure -no-ec2m` correctly leads to "sect163k1" not being available/found (in the groups list in `tls1_group_name2id`), the `s_client` invocation to fail and thus, the issue being impossible to trigger:
```
> LD_LIBRARY_PATH=. ./apps/openssl s_client -no_tls1_3 -groups sect163k1
Error with command: "-groups sect163k1"
```

--> Would it be an (alternative) idea to ensure the same behaviour also in 1.1.1? Requesting an unavailable group _should_ be met with an error message upfront, no?

**Comment by ueno** [comment]
While it seems to be no issue in the master in `-no_tls1.3` case thanks to the `use_ecc` check in `tls_construct_ctos_supported_groups`,  it raises internal error if there are no groups allowed in TLS 1.3:
```console
openssl s_client -cipher DEFAULT:@SECLEVEL=4 -tls1_3 -groups P-256 -port 4443
C0E151C76E7F0000:error:0A0C0103:SSL routines:tls_construct_ctos_supported_groups:internal error:ssl/statem/extensions_clnt.c:236:
```
Perhaps it might make sense to return a meaningful error in that case, e.g., `SSL_R_NO_SHARED_GROUPS`, indicating this is a configuration error as that extension is mandatory in TLS 1.3?

**Comment by mattcaswell** [comment]
With PR #13916 applied the error is different:

````
$ openssl s_client -cipher DEFAULT:@SECLEVEL=4 -tls1_3 -groups "P-256"
CONNECTED(00000003)
80B23F50A17F0000:error:0A000127:SSL routines:tls_construct_ctos_supported_groups:no suitable groups:ssl/statem/extensions_clnt.c:238:No groups enabled for max supported SSL/TLS version
````

**Comment by t8m** [comment]
Fix in #18213

**Comment by t8m** [comment]
Fix was merged.
